### PR TITLE
[Bugfix] break cycles caused by symlinks

### PR DESCRIPTION
**Pull request by stefanpenner**
@rwjblue / @krisselden review?

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -57,7 +59,6 @@ test('walkSync', function (t) {
     'some-other-dir/',
     'some-other-dir/qux.txt',
     'symlink1/',
```
Comment (stefanpenner) [inline test/test.js]: Should we not include `symlink1/` as `some-other-dir` has already been visited?

Comment (stefanpenner) [reply test/test.js]: thinking about this, ya i'll fix.

File: index.js [hunk 2]
```diff
@@ -106,10 +106,29 @@ function _walkSync(baseDir, options, _relativePath) {
...
+
+        var before = results.length;
+
+        if (options.directories !== false && (!globMatcher || globMatcher.match(entry.relativePath))) {
+          results.push(entry);
+        }
+
+        results = results.concat(_walkSync(baseDir, options, entry.relativePath, visited));
+
+        if (results.length === before) {
```
Comment (stefanpenner) [inline index.js]: this part ended up being slightly tricky

File: index.js [hunk 3]
```diff
@@ -41,16 +41,16 @@ function walkSync(baseDir, _options) {
     };
   }
 
-  return _walkSync(baseDir, options).map(mapFunct);
+  return _walkSync(baseDir, options, null, Object.create(null)).map(mapFunct);
```
Comment (rwjblue) [inline index.js]: Just curious, have you thought about a Set here instead (we don’t really need a “value” per-se)? Any insight on why you preferred a `Object.create(null)`?

File: index.js [hunk 4]
```diff
@@ -106,10 +106,29 @@ function _walkSync(baseDir, options, _relativePath) {
     var entry = sortedEntries[i];
 
     if (entry.isDirectory()) {
-      if (options.directories !== false && (!globMatcher || globMatcher.match(entry.relativePath))) {
-        results.push(entry);
+
+      var realPath = fs.realpathSync(baseDir + '/' + entry.relativePath);
```
Comment (rwjblue) [inline index.js]: It’s a bit unfortunate that in order to bust cycles we have to do an extra I/O operation per directory.

File: index.js [hunk 5]
```diff
@@ -106,10 +106,29 @@ function _walkSync(baseDir, options, _relativePath) {
...
+        var before = results.length;
+
+        if (options.directories !== false && (!globMatcher || globMatcher.match(entry.relativePath))) {
+          results.push(entry);
+        }
+
+        results = results.concat(_walkSync(baseDir, options, entry.relativePath, visited));
+
+        if (results.length === before) {
+          // We unvisit this path, as although we did travers the findings were
```
Comment (rwjblue) [inline index.js]:
```suggestion
          // We unvisit this path, as although we did traverse the findings were
```

File: index.js (hunk 3 above)
Comment (stefanpenner) [reply index.js]: this library supports node 0.10 still...

Comment (stefanpenner) [reply index.js]: Ah, node 0.10 has a set..

File: index.js (hunk 4 above)
Comment (stefanpenner) [reply index.js]: ya :(

File: index.js (hunk 3 above)
Comment (rwjblue) [reply index.js]: 😭 so when we drop it, should we experiment with using a Set?

Comment (stefanpenner) [reply index.js]: ya

**Pull request merged**